### PR TITLE
fix(subscriberdb): Handling sqlite lock failure

### PR DESCRIPTION
**Pull request by panyogesh**
Fix:
  1. Handling the OperationalError case for sqlite3
  2. Dumping the process holding the lock
  3. Return the Server too busy error message to AMF in case  if db is locked

 Test:
         1. Simulated the issue with
             connection.isolation_level = 'EXCLUSIVE'
             connection.execute('BEGIN EXCLUSIVE')
          2. Tested with the lock and without lock on simulator

 Logs:
       1. For process holding the lock
           subscriberdb[2785574]:
              root      2784266 F.... (root)python3
         subscriberdb[2785574]:
             root      2785490 F.... (root)python3.8

      Feb  6 19:44:58 magma-dev-focal subscriberdb[2784266]:
         ERROR:root:Sqlite3 DB is locked for 222456000000121:
         IMSI222456000000121

Signed-off-by: Yogesh Pandey <yogesh@wavelabs.ai>

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

**Comment by panyogesh** [comment]
[Test-Logs-Subscriberdb-Lock-Feb-7.txt](https://github.com/magma/magma/files/8011179/Test-Logs-Subscriberdb-Lock-Feb-7.txt)

**Comment by panyogesh** [comment]
7/23 Test  #7: test_itti ..................................   Passed    4.55 sec
      Start  8: test_amf_app

File: lte/gateway/python/magma/subscriberdb/store/sqlite.py [hunk 1]
```diff
@@ -22,7 +22,12 @@
 from magma.subscriberdb.sid import SIDUtils
 from orc8r.protos.digest_pb2 import Digest, LeafDigest
 
-from .base import BaseStore, DuplicateSubscriberError, SubscriberNotFoundError
+from .base import (
```
Comment (ardzoht) [inline lte/gateway/python/magma/subscriberdb/store/sqlite.py]: Let's use absolute imports for all python code

File: lte/gateway/python/magma/subscriberdb/store/sqlite.py [hunk 2]
```diff
@@ -22,7 +22,12 @@
...
+    BaseStore,
+    DuplicateSubscriberError,
+    SubscriberNotFoundError,
+    SubscriberServerTooBusy,
+)
 from .onready import OnDataReady, OnDigestsReady
```
Comment (ardzoht) [inline lte/gateway/python/magma/subscriberdb/store/sqlite.py]: can you also update this import to absolute method?

File: lte/gateway/python/magma/subscriberdb/store/sqlite.py [hunk 3]
```diff
@@ -264,6 +269,14 @@ def get_subscriber_data(self, subscriber_id):
                 row = res.fetchone()
                 if not row:
                     raise SubscriberNotFoundError(subscriber_id)
+        except sqlite3.OperationalError:
+            # Print the process holding the lock
+            db_parts = db_location.split(":", 1)
+            if (len(db_parts) == 2) and db_parts[1]:
+                path_str = db_parts[1].split("?")
+                import os
```
Comment (ardzoht) [inline lte/gateway/python/magma/subscriberdb/store/sqlite.py]: I don't think there's much benefits into using conditional imports here, unless we have any explicit evidence that there could be performance impact, let's keep this import on the top of the file

File: lte/gateway/python/magma/subscriberdb/store/sqlite.py [hunk 4]
```diff
@@ -264,6 +269,14 @@ def get_subscriber_data(self, subscriber_id):
...
+                os.system('fuser -uv ' + path_str[0])
```
Comment (ardzoht) [inline lte/gateway/python/magma/subscriberdb/store/sqlite.py]: update this to `subprocess.call()` or `Popen()`, retrieve the stdout and then log it using `logging` library

**Review by ardzoht — CHANGES_REQUESTED** [review]

**Comment by panyogesh** [comment]
[Test-Logs-Subscriberdb-Lock-Feb-8.txt](https://github.com/magma/magma/files/8022623/Test-Logs-Subscriberdb-Lock-Feb-8.txt)

Attaching the updated logs.

File: lte/gateway/python/magma/subscriberdb/store/sqlite.py (hunk 1 above)
Comment (panyogesh) [reply lte/gateway/python/magma/subscriberdb/store/sqlite.py]: Comments addressed.

Comment (panyogesh) [reply lte/gateway/python/magma/subscriberdb/store/sqlite.py]: Comments addressed.

File: lte/gateway/python/magma/subscriberdb/store/sqlite.py (hunk 2 above)
Comment (panyogesh) [reply lte/gateway/python/magma/subscriberdb/store/sqlite.py]: Comments addressed.

File: lte/gateway/python/magma/subscriberdb/store/sqlite.py (hunk 3 above)
Comment (panyogesh) [reply lte/gateway/python/magma/subscriberdb/store/sqlite.py]: Comments addressed.

File: lte/gateway/python/magma/subscriberdb/store/sqlite.py (hunk 4 above)
Comment (panyogesh) [reply lte/gateway/python/magma/subscriberdb/store/sqlite.py]: Comments addressed.

**Comment by ardzoht** [comment]
@panyogesh How frequent is the issue happening on your testing? Also does this happen with reaching a certain amount of attached subscribers / how does the load look like?

**Review by ardzoht — DISMISSED** [review]
lgtm, left some questions regarding the failure itself

**Review by ardzoht — CHANGES_REQUESTED** [review]
@panyogesh How frequent is the issue happening on your testing? If this is happening really frequently have we tested the performance impact of dropping running processes over time?

**Comment by panyogesh** [comment]
> @panyogesh How frequent is the issue happening on your testing? If this is happening really frequently have we tested the performance impact of dropping running processes over time?

@ardzoht : Interestingly this issue is coming only on one setup from TerravM which is the regression setup. There is no fixed pattern but on an average 2 testcases fails out of 95 cases and these are random failures.  Same script when executed on other TVM setups or in scale test setups (where we run continuous registration/deregistration for 4-5 hours) this issue is not seen. So far we have not done any performance impact as in other setups issue is not reproduced.

**Review by ardzoht — APPROVED** [review]
lgtm

**Pull request merged**